src/transformers/models/llama/tokenization_llama.py

```diff
@@ -0,0 +1,209 @@
+# coding=utf-8
+# Copyright 2022 The FAIR team of Meta AI and The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tokenization classes for LLaMA."""
+import os
+import re
+from shutil import copyfile
+from typing import Any, Dict, List, Optional, Tuple
+
+import sentencepiece as spm
+
+from ...tokenization_utils import PreTrainedTokenizer
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+VOCAB_FILES_NAMES = {"vocab_file": "tokenizer.model"}
+
+PRETRAINED_VOCAB_FILES_MAP = {}
+
+
+class LLaMATokenizer(PreTrainedTokenizer):
+    """
+    Construct a LLaMA tokenizer. Based on byte-level Byte-Pair-Encoding.
+
+    Args:
+        vocab_file (`str`):
+            Path to the vocabulary file.
+    """
+
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+    model_input_names = ["input_ids", "attention_mask"]
+
+    def __init__(
+        self,
+        vocab_file,
+        unk_token="",
+        bos_token="",
+        eos_token="",
```

Comment:
This adds tokens if you choose another str value, which is probably something you don't want, you probably want to remap `bos_token` (whatever is it's "" or "<s>" or "<bos>") to bos_token_id, which is 1 in this case.

Reply:
Sorry, could you describe this in a little more detail?

Reply:
Right! The issue at hand is an interesting one. `bos_token`/`eos_token` are "special" tokens, in the sense that whenever you tokenize a text, they should never spontaneously appear, they should instead be added programmatically at the beginning of each document.

Assume you say `bos_token="<bos>"` and text `This token signals the beginning of a sequence and is represented by "<bos>"`. You somehow don't want to tokenize `<bos>` as the bos_token. It's supposed to be considered as any string, and thus should be represented as a list of token ids, just not `1`.

The current issue is that the current behaviour does this:
```python
>>> from transformers import AutoTokenizer
>>> tokenizer = AutoTokenizer.from_pretrained("/fsx/thomaswang/experiments/llama/hub/llama-7b", add_bos_token=True, bos_token="<bos>")
tokenizer("Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
>>> tokenizer("<bos>")
{'input_ids': [1, 32000], 'attention_mask': [1, 1]}
```

Whereas if you tokenizer "<bos>" with sentencepiece you get:
```python
>>> from sentencepiece import SentencePieceProcessor
>>> tokenizer = SentencePieceProcessor(model_file="tokenizer.model")
>>> tokenizer.encode("<bos>")
[529, 27737, 29958]
```
Let me know if I'm still being unclear.